{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 2: &nbsp; _k_-mer Analysis and de Bruijn Graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Due Friday Apr 29"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If two or more people are working on this project enter each person's name here:\n",
    "* Name:\n",
    "* Name:\n",
    "* Name:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparation "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the paper by Compeau, Pevzner, and Tesler (\"How to apply de Bruijn graphs to genome assembly\")."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this project I recommend `scikit-bio` for its Sequence class and an extension to `scipy` called APGL (Another Python Graph Library).  You are welcome to use alternatives, but if you do make sure you tell me which libraries you used and what I might need to do to install them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from skbio import io, Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from scipy.sparse import lil_matrix\n",
    "from apgl.graph import SparseGraph, GeneralVertexList\n",
    "from collections import Counter\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Execute this code cell to import the PyPlot library and tell Jupyter to insert plots in the notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict \n",
    "\n",
    "% matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Execute these code cells to define the circular \"genome\" and set of five \"reads\" shown in Figure 3 of paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tiny_genome = Sequence('ATGGCGTGCA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tiny_reads = [\n",
    "    Sequence('CGTGCAA'),\n",
    "    Sequence('TGCAATG'),\n",
    "    Sequence('ATGGCGT'),\n",
    "    Sequence('GGCGTGC'),\n",
    "    Sequence('CAATGGC'),\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1: &nbsp; `kmers`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fill in the body of the function named `kmers` so the function returns all substrings of length _k_ found in a set of sequences.  When it is called we'll pass it a series of Sequence objects and an integer `k`; the function should return a dictionary where keys are _k_-mers found in the sequences and the number of times they were found.\n",
    "\n",
    "Example:\n",
    "```\n",
    ">>> kmers(tiny_reads, 3)\n",
    "{'AAT': 2,\n",
    " 'ATG': 3,\n",
    " ...\n",
    " 'TGG': 2}\n",
    "```\n",
    "\n",
    "You can write your own loop or use the `scikit-bio` method named `iter_kmers` to make the _k_-mers. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Documentation and Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "ec32d7e36db2b3f5dd639545a5d2ad29",
     "grade": true,
     "grade_id": "kmer_doc",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "source": [
    "Use a sliding window to produce kmers of length k from the list of reads and then use the Counter object to create a dictionary of kmer counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "nbgrader": {
     "checksum": "9cbf0cc41d641b1fe48f871e1bdadb09",
     "grade": true,
     "grade_id": "kmer",
     "locked": false,
     "points": 10,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "def kmers(reads, k):\n",
    "    \n",
    "    kmers = []\n",
    "    \n",
    "    for i in reads:\n",
    "        for j in range(len(i)+1-k):\n",
    "            kmers.append(str(i[j:j+k]))\n",
    "            \n",
    "    kmer_count = Counter(kmers)\n",
    "    \n",
    "    return kmer_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Autograder Tests "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The notebook grader will run these tests in the order shown.  You are welcome to run these tests, too, but be careful to run them in order (a good set of unit tests should run in any order but some of our tests will use data sets that take several seconds to create so we'll put the \"setup\" step in a separate cell)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tiny_kmers = kmers(tiny_reads, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "7761178696157fb27eb4d330087199a7",
     "grade": true,
     "grade_id": "kmer_is_dict",
     "locked": true,
     "points": 1,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert isinstance(tiny_kmers, dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "dcd26c9f580d50fd1047882daa7a41c5",
     "grade": true,
     "grade_id": "kmer_has_10",
     "locked": true,
     "points": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert len(tiny_kmers) == 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "nbgrader": {
     "checksum": "7842047ae4b4de220c377b0b2d293273",
     "grade": true,
     "grade_id": "kmer_known_3",
     "locked": true,
     "points": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "for kmer in ['ATG', 'TGG', 'CAA']:\n",
    "    assert kmer in tiny_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "f740f7dbab91f17904e9b4bb9bf832d5",
     "grade": true,
     "grade_id": "kmer_sum",
     "locked": true,
     "points": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert sum(tiny_kmers.values()) == 25"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2: &nbsp; _k_-mer Spectrum "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fill in the body of the function named `plot_spectrum` so it draws the _k_-mer spectrum for a set of sequences.  The argument passed to the function will be a dictionary like the one produced by a call to `kmers`, where keys are strings of length _k_ and values are the number of times that string was found.\n",
    "\n",
    "A spectrum is a plot that has _k_-mer frequency on the _x_ axis.  For any point _x_, the _y_ value is the number of _k_-mers that occur _x_ times.\n",
    "\n",
    "The easiest way to create a spectrum plot is to generate a histogram using the set of counts in the _k_-mer dictionary.  You can test your function on the _k_-mers of the \"tiny reads\" data set but the plot will be pretty boring.  5 _k_-mers appear 2 times, and 5 more appear 3 times, so the histogram will have two bars, one at _x_ = 2 and one at _x_ = 3, and both bars will top out at _y_ = 5.  Execute this code cell to see what that might look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<img src=\"http://www.cs.uoregon.edu/Classes/16S/cis454/graphics/lectures/spectrum.svg\"/>"
      ],
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "Image(url='http://www.cs.uoregon.edu/Classes/16S/cis454/graphics/lectures/spectrum.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Documentation and Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "f4bcab4b242bfea797427bea37ee1864",
     "grade": true,
     "grade_id": "spectrum_doc",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "source": [
    "Use histogram function of matplotlib to plot a histogram of kmer frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "nbgrader": {
     "checksum": "e52b92d13e03585587eb5f4295fdca1b",
     "grade": true,
     "grade_id": "spectrum",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "def plot_spectrum(kmers):\n",
    "    \n",
    "    frequency = sorted(list(kmers.values()))\n",
    "    \n",
    "    plt.hist(frequency, alpha = 1)\n",
    "    plt.margins(0.1)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWkAAAD9CAYAAAB++4avAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAADSFJREFUeJzt3W+M5HddwPH3dad3unfLdc9OCwixsXofCWJ5AFGatpAK\nDxBKqqJEPbVUbZHGFDEmQlNNNGBNlWioCrSllT4CTAlilKqxGEQDgokhop/rHwHjn3Tpjte7bnu9\n3q0PZhfX5XZmujvznc9s369HNze/7vfzvd/ee3/7m53rntXVVSRJNZ0z7QEkSVsz0pJUmJGWpMKM\ntCQVZqQlqTAjLUmFdcb9AZeWjjf9mb7FxXl6vZWWSza1m/e3m/cG7m/Wtd5ft7uw52y/P/NX0p3O\n3LRHmKjdvL/dvDdwf7Ouyv5mPtKStJsZaUkqzEhLUmFGWpIKM9KSVJiRlqTCjLQkFWakJakwIy1J\nhRlpSSrMSEtSYUZakgob6V/Bi4gvAMfWHv5bZv7M5EaSJK0bGumI2AeQmVdOfhxJ0kajXElfAuyP\niPuAOeCmzPzsZMeSJMFokV4Bbs3MOyPiO4E/j4jDmXlmwrNJEgCnT5/my19+uOmahw5d0nS9rYwS\n6aPAgwCZ+UBEPAo8D/iPsx28uDjf7B/Lvu0D9/AXn/v3JmsBHNr3BHf//m80W29dt7vQfM1WdvPe\nwP2Ny9GjR7nx1j9h/uAFTdZbOfYI9/zmAQ4fPtxkvUFGifS1wEuAGyLi+cAC8F9bHdzyfzdz7PhJ\nVg++uNl6q08/yNLS8WbrQf8vQes1W9nNewP3N07LyyeYP3gBBxa/tcl661qev62+4I0S6TuBuyLi\n08AZ4FpvdUhSG0MjnZmngCMNZpEkbeKbWSSpMCMtSYUZaUkqzEhLUmFGWpIKM9KSVJiRlqTCjLQk\nFWakJakwIy1JhRlpSSrMSEtSYUZakgoz0pJUmJGWpMKMtCQVZqQlqTAjLUmFGWlJKsxIS1JhRlqS\nCjPSklSYkZakwoy0JBVmpCWpMCMtSYUZaUkqzEhLUmFGWpIKM9KSVFhnlIMi4gLg88CrM/PoZEeS\nJK0beiUdER3gfcDK5MeRJG00yu2O3wb+EPjPCc8iSdpk4O2OiLgGeCQz/zIi3jnKB1xcnKfTmRvH\nbEPtn9/XZJ115+6do9tdaLomMJU1W9nNewP3Ny693oEm62xW4fwNuyf9ZuBMRLwGeCnwoYh4Q2Y+\nstV/0Ou1uyvy+MrJZmsBnHrqNEtLx5uu2e0uNF+zld28N3B/47S8fKLJOpu1PH9bfUEYGOnMfOX6\nryPifuD6QYGWJI3XM/kRvNWJTSFJOquRfgQPIDOvnOQgkqRv5JtZJKkwIy1JhRlpSSrMSEtSYUZa\nkgoz0pJUmJGWpMKMtCQVZqQlqTAjLUmFGWlJKsxIS1JhRlqSCjPSklSYkZakwoy0JBVmpCWpMCMt\nSYUZaUkqzEhLUmFGWpIKM9KSVJiRlqTCjLQkFWakJakwIy1JhRlpSSrMSEtSYUZakgrrDDsgIs4B\nbgcCOAO8JTO/NOnBJEmjXUlfBaxm5mXAzcC7JzuSJGnd0Ehn5seB69YeXgT0JjmQJOn/DL3dAZCZ\nZyLibuBq4I0TnUiS9HUjRRogM6+JiAuAz0XEizLzibMdt7g4T6czN7YBB9k/v6/JOuvO3TtHt7vQ\ndE1gKmu2spv3Bu5vXHq9A03W2azC+RvlhcMjwAsy8xbgSeA0/RcQz6rXWxnfdEM8vnKy2VoAp546\nzdLS8aZrdrsLzddsZTfvDdzfOC0vn2iyzmYtz99WXxBGuZK+F7grIv5m7fgbM7NtHSXpWWpopDNz\nBXhTg1kkSZv4ZhZJKsxIS1JhRlqSCjPSklSYkZakwoy0JBVmpCWpMCMtSYUZaUkqzEhLUmFGWpIK\nM9KSVJiRlqTCjLQkFWakJakwIy1JhRlpSSrMSEtSYUZakgoz0pJUmJGWpMKMtCQVZqQlqTAjLUmF\nGWlJKsxIS1JhRlqSCjPSklSYkZakwjqDnoyIDvBB4CJgL/CuzPxEg7kkSQy/kj4CfC0zrwBeC9w2\n+ZEkSesGXkkDHwE+uvbrc4BTkx1HkrTRwEhn5gpARCzQj/VNLYaSJPUNu5ImIl4I3AvclpkfHnb8\n4uI8nc7cOGYbav/8vibrrDt37xzd7kLTNYGprNnKbt4buL9x6fUONFlnswrnb9gLhxcC9wE3ZOb9\no3zAXm9lHHON5PGVk83WAjj11GmWlo43XbPbXWi+Ziu7eW/g/sZpeflEk3U2a3n+tvqCMOxK+h3A\necDNEfGrwCrw2sxsW0dJepYadk/6bcDbGs0iSdrEN7NIUmFGWpIKM9KSVJiRlqTCjLQkFWakJakw\nIy1JhRlpSSrMSEtSYUZakgoz0pJUmJGWpMKMtCQVZqQlqTAjLUmFGWlJKsxIS1JhRlqSCjPSklSY\nkZakwoy0JBVmpCWpMCMtSYUZaUkqzEhLUmFGWpIKM9KSVJiRlqTCjLQkFTZSpCPieyPi/kkPI0n6\n/zrDDoiIXwZ+Ejgx+XEkSRuNciX9IPCDkx5EkvSNhkY6Mz8GPN1gFknSJkNvdzxTi4vzdDpz4/6w\nZ7V/fl+Tddadu3eObneh6ZrAVNZsZTfvDdzfuPR6B5qss1mF8/dMIr1nlIN6vZVtjvLMPb5ystla\nAKeeOs3S0vGma3a7C83XbGU37w3c3zgtL0/nJbGW52+rLwjP5EfwVscziiRpVCNdSWfmV4BLJzyL\nJGkT38wiSYUZaUkqzEhLUmFGWpIKM9KSVJiRlqTCjLQkFWakJakwIy1JhRlpSSrMSEtSYUZakgoz\n0pJUmJGWpMKMtCQVZqQlqTAjLUmFGWlJKsxIS1JhRlqSCjPSklSYkZakwoy0JBVmpCWpMCMtSYUZ\naUkqzEhLUmFGWpIKM9KSVFhn2AERsQf4A+AS4EngZzPz4UkPJkka7Ur6amBfZl4KvAN4z2RHkiSt\nGyXSlwGfBMjMzwIvm+hEkqSvG3q7A3gOcGzD46cj4pzMPDOhmUp76KEHmq7X6x1geflE0zVb2c17\nA/c3Tl/96ldYOfZIk7WApmsNM0qkHwMWNjweGOjFxXk6nbkdDzaKq3/gcs7/+39sshbA8y78Hg4d\nOtBsvXXTWLOV3bw3cH/jcvDgd3HPS1/cZK11F198MXNzbVo2yCiR/gzweuCPI+L7gC8OOrjXWxnH\nXCO54Pznc/21wdLS8WZrttbtLuza/e3mvYH7m3Vzc3NN99ftLpz190eJ9MeA10TEZ9Yev3lcQ0mS\nBhsa6cxcBX6+wSySpE18M4skFWakJakwIy1JhRlpSSrMSEtSYUZakgoz0pJUmJGWpMKMtCQVZqQl\nqTAjLUmFGWlJKsxIS1JhRlqSCjPSklTYntXV1WnPIEnaglfSklSYkZakwoy0JBVmpCWpMCMtSYUZ\naUkqrDPtAUYRER3gg8BFwF7gXZn5iQ3PXwXcDJwC7srMO6Yx53aNsL8fA26kv78vZuZbpzHndg3b\n34bj3g88mpnvbDvh9o1w7l4O/M7aw/8GjmTmU63n3K4R9vcTwNuBp+n/3XvfNObcrog4B7gdCOAM\n8JbM/NKG56fellm5kj4CfC0zrwBeC9y2/sTaJ9F7gFcDrwKui4juNIbcgUH7+ybg14FXZublwHkR\n8frpjLltW+5vXURcD3x368HGYNjePgBcs/b8J4FvazzfTg3b363AlcBlwC9FxMHG8+3UVcBqZl5G\nP8bvXn+iSltmJdIfof8HCP2ZT2147kXAA5n5WGaeAv4WuKLxfDs1aH8ngUsz8+Ta4w7wZMPZxmHQ\n/oiIVwAvB97feK5x2HJvEXEYeBR4e0R8CjiUmQ80n3BnBp474J+AReCb1x7P1LvjMvPjwHVrDy8C\nehueLtGWmbjdkZkrABGxAHwUuGnD088Bjm14fByYqa/mg/aXmavA0trzvwDsz8y/msac2zVofxHx\nXODXgKuBN01lwB0Y8rl5PvAK4K3Aw8CfRsTnM/NTrefcriH7A/hn4AvACeDezHys7YQ7l5lnIuJu\n+p+Db9zwVIm2zMqVNBHxQuCvgT/KzA9veOox+n+Y6xaA/2k52zgM2B8RsScibgW+H/ihacy3UwP2\n9yPAtwB/BvwK8OMR8VNTGHHbBuztUeDBzDyamU/Tv93xsmnMuBNb7S8iXgK8jv4tnIuACyPih6cy\n5A5l5jXAYeCOiFj/rqBEW2biSjoiLgTuA27IzPs3Pf0vwHdExHnACv1vR25tPOKODNkf9O9rPpGZ\nV7edbDwG7S8z3wu8d+24nwYiMz/UfsrtGXLuHgYORMS3Z+bDwOXArL2oPWh/x+j/nTuZmasR8Qj9\nWx8zIyKOAC/IzFvo30Y8Tf8FRCjSlpn4B5Yi4neBHwX+FdhD/77X7fS/9b8jIl5H/1vmPcCdM/gK\n85b7o/+t5D8An147fBX4vbV7aTNh2PnbcNx6pGfppzuGfW6+CvittcP/LjN/cSqDbtMI+7seuJb+\naycPAT+39l3DTIiIeeAu4Ln0L1pvAQ5QqC0zEWlJeraamXvSkvRsZKQlqTAjLUmFGWlJKsxIS1Jh\nRlqSCjPSklSYkZakwv4XN/CB6dawkJUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x110baca20>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_spectrum(tiny_kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 3: &nbsp; Plasmid Reads "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get a better idea of what the _k_-mer spectrum from a set of Illumina reads looks like download the file named `ColE9-J.fastq` from Canvas.  This file has 2,500 reads, each 150bp, from an _E. coli_ plasmid named ColE9-J. \n",
    "\n",
    "For this part of the project you will generate a _k_-mer spectrum plot for this data set, using a value of _k_ that might be used by a genome assembler.\n",
    "\n",
    "Execute this code cell to make a list containing all 2,500 reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plasmid_reads = list(io.read('ColE9-J.fastq', format='fastq', variant='illumina1.8'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As an additional test of your `plot_spectrum` function I suggest making a plot that uses only a few of the reads (it takes over a minute to count _k_-mers of the full data set on my machine).  This code cell shows how to get 21-mer counts using the first 100 reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plasmid_kmers = kmers(plasmid_reads[:100], 21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD9CAYAAACyYrxEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEFRJREFUeJzt3W+MXFd9xvGvvWPHazI2S1lHlH8Gk/5SqVLaUKUNNXZA\nsUgiILQvUhVBAxIJRFFKWhU1CTISbU1Qk6ZNSglSYkhIqrZgQYFGMaUEyDpBJWmpUov0h0tkLJWq\nXuKN/7Brx7vevpixOh52dwZ79u7M8ffzxtlzz97z3BPnmcm9s/ay2dlZJEnlWr7UASRJi8uil6TC\nWfSSVDiLXpIKZ9FLUuEsekkqXG2pA8xlfPxwpZ/5HBlZzcTEZJVL9sSg5obBzT6ouWFwsw9qbqg+\n++hofdlc476jB2q1oaWOcFoGNTcMbvZBzQ2Dm31Qc0P/ZLfoJalwFr0kFc6il6TCWfSSVDiLXpIK\nZ9FLUuEsekkqnEUvSYWz6CWpcBa9JBXOopekwln0klQ4i16SCmfRS1LhLHpJKpxFL0mF68u/YapK\nd3xiO8/89yQz0ycqWe/VL13OR//oxkrWkiSw6GHoHGbXvqay/7UZWrmvopUkqcFbN5JUOItekgpn\n0UtS4Sx6SSqcRS9JhbPoJalwFr0kFc6il6TCWfSSVLiOPxkbETXgAWA9MA1cC8wA9wMngN2ZeUNz\n7rXAdcBxYFtmPhwRq4CHgHXAIeCazHyu51ciSZpTN+/orwSGMvM3gD8BPgbcCdyamZuB5RFxVUSc\nB9wIXAJcDtwWESuA64GnM3MT8CCwdRGuQ5I0j26K/vtALSKWAWtpvFu/KDPHmscfAbYAFwO7MnM6\nMw8Be4ALgY3Azpa5l/UwvySpg27+ULMjwGuA/wR+Dngb8MaW44eBNUAdONj2fWvbxk/OlSRVpJui\n/31gZ2Z+OCJeDnwTWNlyvA48T+P++5q28YnmeL1t7oJGRlZTqw11Ee3MDQ+vOPXlaZGds6rG6Gi9\n88Qu9fJcVRvU7IOaGwY3+6Dmhv7I3k3RH6BxuwYaJV0DvhsRmzPzW8AVwKPAk8C2iFgJDAMXALuB\nJ2jc53+q+esYHUxMTP6Ml3H6pqaOd57UQ8eOTjM+frgn5xodrffsXFUb1OyDmhsGN/ug5obqs8/3\notJN0f8l8OmIeAxYAdwM/CtwX/Nh6zPAjsycjYi7gV3AMhoPa1+IiHuAByJiDDgGvPOMr0aS1LWO\nRZ+ZPwF+e45Dl84xdzuwvW1sCrj6NPNJks6QPzAlSYWz6CWpcBa9JBXOopekwln0klQ4i16SCmfR\nS1LhLHpJKpxFL0mFs+glqXAWvSQVzqKXpMJZ9JJUOItekgpn0UtS4Sx6SSqcRS9JhbPoJalwFr0k\nFc6il6TCWfSSVDiLXpIKZ9FLUuEsekkqnEUvSYWz6CWpcBa9JBXOopekwln0klQ4i16SCmfRS1Lh\nLHpJKpxFL0mFs+glqXAWvSQVrtbNpIi4GXg7sAL4JPAYcD9wAtidmTc0510LXAccB7Zl5sMRsQp4\nCFgHHAKuycznenwdkqR5dHxHHxGbgUsy8w3ApcCrgDuBWzNzM7A8Iq6KiPOAG4FLgMuB2yJiBXA9\n8HRmbgIeBLYuypVIkubUza2btwC7I+IfgC8D/whclJljzeOPAFuAi4FdmTmdmYeAPcCFwEZgZ8vc\ny3qYX5LUQTe3bl5K4138W4HX0ij71heIw8AaoA4cbBk/AqxtGz85V5JUkW6K/jngmcycBr4fEUeB\nV7QcrwPP07j/vqZtfKI5Xm+bu6CRkdXUakNdRDtzw8MrTn15WmTnrKoxOlrvPLFLvTxX1QY1+6Dm\nhsHNPqi5oT+yd1P0u4DfA/4iIn4eeBHw9YjYnJnfAq4AHgWeBLZFxEpgGLgA2A08AVwJPNX8deyn\nlzjVxMTkaVzK6ZmaOl7ZWgDHjk4zPn64J+caHa337FxVG9Tsg5obBjf7oOaG6rPP96LSseibn5x5\nY0R8B1hG4+HqXuC+5sPWZ4AdmTkbEXfTeGFYRuNh7QsRcQ/wQESMAceAd/bigiRJ3enq45WZefMc\nw5fOMW87sL1tbAq4+nTCSZLOnD8wJUmFs+glqXAWvSQVzqKXpMJZ9JJUOItekgpn0UtS4Sx6SSqc\nRS9JhbPoJalwFr0kFc6il6TCWfSSVDiLXpIKZ9FLUuEsekkqnEUvSYWz6CWpcBa9JBXOopekwln0\nklQ4i16SCmfRS1LhLHpJKpxFL0mFs+glqXAWvSQVzqKXpMJZ9JJUOItekgpn0UtS4Sx6SSqcRS9J\nhbPoJalwFr0kFa7WzaSIWAc8BVwGzAD3AyeA3Zl5Q3POtcB1wHFgW2Y+HBGrgIeAdcAh4JrMfK7X\nFyFJml/Hd/QRUQM+BUw2h+4Ebs3MzcDyiLgqIs4DbgQuAS4HbouIFcD1wNOZuQl4ENi6CNcgSVpA\nN7du7gDuAX4ELAMuysyx5rFHgC3AxcCuzJzOzEPAHuBCYCOws2XuZT3MLknqwoJFHxHvAfZn5tdo\nlHz79xwG1gB14GDL+BFgbdv4ybmSpAp1ukf/XuBERGyh8Q79s8Boy/E68DyN++9r2sYnmuP1trkd\njYysplYb6mbqGRseXnHqS9QiO2dVjdHReueJXerluao2qNkHNTcMbvZBzQ39kX3Bom/ehwcgIh4F\nPgDcHhGbMvMx4ArgUeBJYFtErASGgQuA3cATwJU0HuReCYzRhYmJyc6TemRq6nhlawEcOzrN+Pjh\nnpxrdLTes3NVbVCzD2puGNzsg5obqs8+34vK6Xy88g+BP46Ix4EVwI7M/F/gbmAX8M80Hta+QOPe\n/i9FxBjwPuCjp7GeJOkMdPXxSoDMfHPLl5fOcXw7sL1tbAq4+nTDSZLOnD8wJUmFs+glqXAWvSQV\nzqKXpMJZ9JJUOItekgpn0UtS4Sx6SSqcRS9JhbPoJalwFr0kFc6il6TCWfSSVDiLXpIKZ9FLUuEs\nekkqnEUvSYWz6CWpcBa9JBXOopekwln0klQ4i16SCmfRS1LhLHpJKpxFL0mFs+glqXAWvSQVzqKX\npMJZ9JJUOItekgpn0UtS4Sx6SSqcRS9JhbPoJalwtYUORkQN+DSwHlgJbAO+B9wPnAB2Z+YNzbnX\nAtcBx4FtmflwRKwCHgLWAYeAazLzuUW5EknSnDq9o38X8OPM3ARcDnwCuBO4NTM3A8sj4qqIOA+4\nEbikOe+2iFgBXA883fz+B4Gti3QdkqR5dCr6z/H/5TwETAMXZeZYc+wRYAtwMbArM6cz8xCwB7gQ\n2AjsbJl7WQ+zS5K6sOCtm8ycBIiIOvB54MPAHS1TDgNrgDpwsGX8CLC2bfzkXElShRYseoCIeCXw\nBeATmfl3EfFnLYfrwPM07r+vaRufaI7X2+Z2NDKymlptqJupZ2x4eMWpL1GL7JxVNUZH650ndqmX\n56raoGYf1NwwuNkHNTf0R/ZOD2PPA74K3JCZ32gOfzciNmXmY8AVwKPAk8C2iFgJDAMXALuBJ4Ar\ngaeav47RhYmJydO4lNMzNXW8srUAjh2dZnz8cE/ONTpa79m5qjao2Qc1Nwxu9kHNDdVnn+9FpdM7\n+luAFwNbI+IjwCzwQeCvmg9bnwF2ZOZsRNwN7AKW0XhY+0JE3AM8EBFjwDHgnT25GklS1zrdo78J\nuGmOQ5fOMXc7sL1tbAq4+gzySZLOkD8wJUmFs+glqXAWvSQVruPHK6VuzczMsHfvs13NnZg4lwMH\njpzxmuvXv5ahoWo+iisNKotePbN377N88PYvs3rtukrWmzy4n7s+9HY2bDi/kvWkQWXRq6dWr13H\nuSMvX+oYklp4j16SCmfRS1LhLHpJKpxFL0mFs+glqXAWvSQVzqKXpMJZ9JJUOItekgpn0UtS4Sx6\nSSqcRS9JhbPoJalwFr0kFc6il6TCWfSSVDiLXpIKZ9FLUuEsekkqnEUvSYWz6CWpcBa9JBXOopek\nwln0klQ4i16SCmfRS1LhaksdQNLcZmZm2Lv32Z6db2LiXA4cOLLgnPXrX8vQ0FDP1lR/sOilPrV3\n77N88PYvs3rtukrWmzy4n7s+9HY2bDi/kvVUHYte6mOr167j3JGXL3UMDbhFL/qIWAZ8ErgQOAq8\nLzN79/+jkqQFVfEw9h3AOZn5BuAW4M4K1pQkNVVx62YjsBMgM/8lIn61gjUl6RS9frjdjZe85MJK\n15tPFUW/BjjY8vV0RCzPzBMVrN2XfvCDPT05TzefoqjSvn0/ZPLg/srWmzy4n337fljZelDtnp8N\n+9mNXu35vn0/5E/v/Rqrzn1JD1J1dvTIAXb89U2MjLyskvUWsmx2dnZRF4iIPwe+nZk7ml/vy8xX\nLfQ909Mzs7VaNR/x+vZ3/o2nd2clawGcv+HVvOJlL61svSrNzMxU/tG8pVizKu5nby3FtW3YsKHq\nNZfNNVjFO/rHgbcCOyLi14H/6PQNExOTix7qpNe95nwuufgixscPV7Zmr4yO1gcyNwxu9kHNDYOb\nfVBzAwwNDVWafXS0Pud4FUX/RWBLRDze/Pq9FawpSWpa9KLPzFng+sVeR5I0N/+sG0kqnEUvSYWz\n6CWpcBa9JBXOopekwln0klQ4i16SCmfRS1LhLHpJKpxFL0mFs+glqXAWvSQVzqKXpMJZ9JJUOIte\nkgq36H+VoCRpafmOXpIKZ9FLUuEsekkqnEUvSYWz6CWpcBa9JBWuttQBqhYRvwZ8PDPf1Db+NmAr\ncBz4TGbetxT55rNA7puA9wH7m0Pvz8w9VeebS0TUgE8D64GVwLbM/ErL8b7c8y5y9/OeLwfuBQI4\nAXwgM7/Xcrwv9xy6yt63+w4QEeuAp4DLMvP7LeNLvudnVdFHxIeAdwNH2sZrwJ3A64Ep4PGI+FJm\njlef8qfNl7vp9cC7M/O71abqyruAH2fm70bECPDvwFeg7/d83txN/bznbwNmM3NjRGwGPga8A/p+\nz2GB7E19u+/Nvf0UMDnH+JLv+dl26+a/gN+cY/wXgT2ZeSgzjwO7gE2VJlvYfLmh8RvologYi4ib\nK8zUjc/ReCcDjd9rx1uO9fOeL5Qb+njPM/NLwHXNL9cDEy2H+3nPO2WHPt534A7gHuBHbeN9sedn\nVdFn5heB6TkOrQEOtnx9GFhbSaguLJAb4G+BDwBvAjZGxJWVBesgMycz8ycRUQc+D3y45XDf7nmH\n3NDHew6QmSci4n7gLuBvWg717Z6ftEB26NN9j4j3APsz82vAsrbDfbHnZ1XRL+AQjX8hJ9WB55co\ny8/qrsw8kJnTwMPAryx1oFYR8UrgUeCBzPz7lkN9vecL5IY+33OAzHwP8AvAfREx3Bzu6z0/aZ7s\n0L/7/l5gS0R8A/hl4LPN+/XQJ3t+Vt2jb9H+qvsM8LqIeDGNe2ybgNsrT9XZKbkjYg2wOyIuoHH/\n783A9qUINpeIOA/4KnBDZn6j7XDf7vlCuQdgz98FvCIzPw4cBWZoPNiEPt5zWDh7P+97Zm4++c/N\nsn9/Zp58YNwXe362Fv0sQET8DvCizLwvIv4A+CcaZXpfZv7PUgacx1y5bwG+SeM/jK9n5s4lzNfu\nFuDFwNaI+AiN/PfS/3veKXc/7/kXgM9ExLdo/Pd9E/BbEdHvew6ds/fzvp/Ul93in14pSYXzHr0k\nFc6il6TCWfSSVDiLXpIKZ9FLUuEsekkqnEUvSYWz6CWpcP8HC3QaGWPuS1AAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x112f5a550>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_spectrum(plasmid_kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Discussion "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you are confident your _k_-mer counting and plotting functions are working as expected get a count of all 33-mers in the full data set and then make a spectrum plot.\n",
    "\n",
    "A crucial detail about this data set is that it comes from an experiment with **50x coverage**.  That means each fragment of the genome is expected to occur 50 times in the set of reads.\n",
    "\n",
    "Use the markdown cell below to describe your plot.  Some things you might include in your discussion:\n",
    "* can you explain the general shape of the distribution?  \n",
    "* what is the range of counts along the _x_-axis?\n",
    "* why are there so many _k_-mers with small counts?\n",
    "* why is there a maximum in the middle of the range?\n",
    "\n",
    "**Note:** &nbsp; Your discussion should be more than just a short answer to these 4 questions.  I want you to look at the distribution, consider why it has the shape it does, and what that implies about the data set.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "c920b28cea0bb3720a031422c0386fc1",
     "grade": true,
     "grade_id": "discussion",
     "locked": false,
     "points": 10,
     "solution": true
    }
   },
   "source": [
    "General plot shape is kind of Gaussian, with a massive outlier of small k-mer counts. The high amount of coverage gives the distributation shape, so while none of the kmers had the same coverage as the fragments did, quite a few still had high coverage up to 30. This also explains the x-axis, from 0-30, close to the 50x coverage of the genome fragments. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "f40d1d4acab824b2b3634758ad24f8c1",
     "grade": true,
     "grade_id": "plasmid_kmers",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAD9CAYAAABTJWtQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFahJREFUeJzt3X+MXfV55/G37cEGJ9eOYQfUH0mh0fZh/6ickggDa2wI\nWPwIKc3+QSiKlqZbWJBF00hhN3FqxNK4pJuFUjeNqwWnsFDUbVBpq7D82mUXxiEKJo1ErNAnTlmn\nWbVNJzD2XDDYjGf2j3v4cnMznh/Hd+beGb9fEpp7v+d7z3kezvh87jnnzsySiYkJJEkCWNrrAiRJ\n/cNQkCQVhoIkqTAUJEmFoSBJKgwFSVIx0OsCjtXwcLMrn6lds2YlIyMHu7GqnrKP/rEYegD76Dfd\n6mNwsLFksnHPFCoDA8t6XUJX2Ef/WAw9gH30m7nuw1CQJBWGgiSpMBQkSYWhIEkqDAVJUmEoSJIK\nQ0GSVBgKkqRiRj/RHBHrgM9n5oUR8T5gOzAGHAL+bWYOR8R1wPXAm8C2zHwkIk4EHgBOBUaBazPz\n5Yg4B7irmvtkZt5WbecW4EPV+Cczc3c3m5UkTW3aM4WIuBm4G1hRDd0FbM7MDwIPA/8xIk4DbgLO\nBS4Fbo+IE4AbgRcycwNwP7C1WscO4OrMPB9YFxFrI+KXgA2ZuQ74VeCPutWkJGlmZnL56HvAR9qe\nfzQzv109HgDeAM4GdmXmWGaOAnuBtcB64LFq7qPARRHRAJZn5r5q/HFgUzX3CYDM/AGwLCJOqduY\nJGn2pg2FzHyY1qWit57/ECAizgM2A78PrAIOtL3sVWA10Ggbb7aNjbbNbU4yt30dkqR5Uuu3pEbE\nR4HPAJdX9whGaQXDWxrACK2Df6NtbD+tEJhs7uG2ue3zp7Rmzcqu/YKowcHG9JMWAPvoH4uhB7CP\nfjOXfcw6FCLiY7RuKF+QmW8dtJ8DPhcRy4GTgDOBPcCzwOXA89XXocxsRsShiDgD2AdcAtwKHAF+\nLyLuAN4NLMnMV6arp1u/CndwsMHwcLMr6+ol++gfi6EHsI9+060+jhYsswqFiFgK/AHwfeDhiJgA\nns7M/xQR24FdwBJgS2YejogdwH0RMUTrk0rXVKu6AXiQ1uWrJ976lFE17+vVOjbPrkVJ0rFaMjHR\nlb9R0zPd+iM7vovoL4uhj8XQA9hHv+nimcKkf2Rnwf/ltfn0V1/9Hzz89N55296SQ8P8yfbPzdv2\nJMlQmIXX3zjM0lN+cd62NzD6nXnbliSBv+ZCktTGUJAkFYaCJKkwFCRJhaEgSSoMBUlSYShIkgpD\nQZJUGAqSpMJQkCQVhoIkqTAUJEmFoSBJKgwFSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkSYWh\nIEkqDAVJUjEwk0kRsQ74fGZeGBHvBe4FxoE9mbm5mnMdcD3wJrAtMx+JiBOBB4BTgVHg2sx8OSLO\nAe6q5j6ZmbdV67gF+FA1/snM3N29ViVJ05n2TCEibgbuBlZUQ3cCWzJzI7A0Iq6MiNOAm4BzgUuB\n2yPiBOBG4IXM3ADcD2yt1rEDuDozzwfWRcTaiPglYENmrgN+FfijrnUpSZqRmVw++h7wkbbn78/M\noerxo8Am4GxgV2aOZeYosBdYC6wHHmube1FENIDlmbmvGn+8Wsd64AmAzPwBsCwiTqnbmCRp9qYN\nhcx8GBhrG1rS9rgJrAIawIG28VeB1R3jzbax0Y51dM5tX4ckaZ7M6J5Ch/G2xw1gP62D/KqO8ZFq\nvNExt3mUuYfb5rbPn9KaNSsZGFg2uw6OYnCwMeXyRmPFlMu7bWBg6bQ1TabOa/rRYuhjMfQA9tFv\n5rKPOqHwNxGxITOfAS4DngJ2A9siYjlwEnAmsAd4FrgceL76OpSZzYg4FBFnAPuAS4BbgSPA70XE\nHcC7gSWZ+cp0xYyMHKzRwk8aHGwwPNycck6zeYi3b63MvbGx8Wlr6jSTPhaCxdDHYugB7KPfdKuP\nowVLnVD4FHB3dSP5ReChzJyIiO3ALlqXl7Zk5uGI2AHcFxFDwCHgmmodNwAP0rp89cRbnzKq5n29\nWsfmGrVJko7BjEIhM78PnFc93gtcMMmcncDOjrHXgasmmfscrU8qdY7fBtw2k5okSd3nD69JkgpD\nQZJUGAqSpMJQkCQVhoIkqTAUJEmFoSBJKgwFSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkSYWh\nIEkqDAVJUmEoSJIKQ0GSVBgKkqTCUJAkFYaCJKkwFCRJhaEgSSoMBUlSMVDnRRExANwHnA6MAdcB\nR4B7gXFgT2ZuruZeB1wPvAlsy8xHIuJE4AHgVGAUuDYzX46Ic4C7qrlPZuZt9VuTJM1W3TOFy4Fl\nmfmvgd8Bfhe4E9iSmRuBpRFxZUScBtwEnAtcCtweEScANwIvZOYG4H5ga7XeHcDVmXk+sC4i1tZt\nTJI0e3VD4bvAQEQsAVbTemd/VmYOVcsfBTYBZwO7MnMsM0eBvcBaYD3wWNvciyKiASzPzH3V+OPA\nxTXrkyTVUOvyEfAqcAbwt8ApwIeB89uWN4FVQAM40PG61R3jzbax0Y51nDFdIWvWrGRgYFmtJjoN\nDjamXN5orOjKdmZqYGDptDVNps5r+tFi6GMx9AD20W/mso+6ofBJ4LHM/GxE/Azwf4DlbcsbwH5a\nB/lVHeMj1XijY25zkrn7pytkZORgvQ46DA42GB5uTjmn2TwEzF8wjI2NT1tTp5n0sRAshj4WQw9g\nH/2mW30cLVjqXj56hbff6e+nFS7fioiN1dhlwBCwG1gfEcsjYjVwJrAHeJbWfQmqr0OZ2QQORcQZ\n1WWpS6p1SJLmSd0zhbuAL0fEM8AJwKeBbwL3VDeSXwQeysyJiNgO7AKW0LoRfTgidgD3RcQQcAi4\nplrvDcCDtMLqiczcXbcxSdLs1QqFzHwN+Ogkiy6YZO5OYGfH2OvAVZPMfY7WJ5UkST3gD69JkgpD\nQZJUGAqSpMJQkCQVhoIkqTAUJEmFoSBJKgwFSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkSYWh\nIEkqDAVJUmEoSJIKQ0GSVBgKkqTCUJAkFYaCJKkwFCRJhaEgSSoG6r4wIj4N/DJwAvAl4BngXmAc\n2JOZm6t51wHXA28C2zLzkYg4EXgAOBUYBa7NzJcj4hzgrmruk5l5W936JEmzV+tMISI2Audm5nnA\nBcB7gDuBLZm5EVgaEVdGxGnATcC5wKXA7RFxAnAj8EJmbgDuB7ZWq94BXJ2Z5wPrImJt/dYkSbNV\n9/LRJcCeiPhL4K+BrwJnZeZQtfxRYBNwNrArM8cycxTYC6wF1gOPtc29KCIawPLM3FeNPw5cXLM+\nSVINdS8f/QtaZwdXAD9PKxjaA6YJrAIawIG28VeB1R3jzbax0Y51nFGzPklSDXVD4WXgxcwcA74b\nEW8AP9u2vAHsp3WQX9UxPlKNNzrmNieZu3+6QtasWcnAwLKabfy4wcHGlMsbjRVd2c5MDQwsnbam\nydR5TT9aDH0shh7APvrNXPZRNxR2Ab8J/H5E/DTwDuB/RcTGzHwauAx4CtgNbIuI5cBJwJnAHuBZ\n4HLg+errUGY2I+JQRJwB7KN1ierW6QoZGTlYs4UfNzjYYHi4OeWcZvMQMH/BMDY2Pm1NnWbSx0Kw\nGPpYDD2AffSbbvVxtGCpFQrVJ4jOj4jngCW0bhzvA+6pbiS/CDyUmRMRsZ1WiCyhdSP6cETsAO6L\niCHgEHBNteobgAdpXYp6IjN316lPklRP7Y+kZuanJxm+YJJ5O4GdHWOvA1dNMvc5Wp9UkiT1gD+8\nJkkqDAVJUmEoSJIKQ0GSVBgKkqTCUJAkFYaCJKkwFCRJhaEgSSoMBUlSYShIkgpDQZJUGAqSpMJQ\nkCQVhoIkqTAUJEmFoSBJKgwFSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkScXAsbw4Ik4Fngcu\nBo4A9wLjwJ7M3FzNuQ64HngT2JaZj0TEicADwKnAKHBtZr4cEecAd1Vzn8zM246lPknS7NQ+U4iI\nAeCPgYPV0J3AlszcCCyNiCsj4jTgJuBc4FLg9og4AbgReCEzNwD3A1urdewArs7M84F1EbG2bn2S\npNk7lstH/4XWQfwfgCXAWZk5VC17FNgEnA3sysyxzBwF9gJrgfXAY21zL4qIBrA8M/dV44/TOgOR\nJM2TWqEQEb8G/HNmPkkrEDrX1QRWAQ3gQNv4q8DqjvFm29hoxzpW16lPklRP3XsKHwfGI2ITrXf+\n/w0YbFveAPbTOsiv6hgfqcYbHXObk8zdP10ha9asZGBgWb0uOgwONqZc3mis6Mp2ZmpgYOm0NU2m\nzmv60WLoYzH0APbRb+ayj1qhUN03ACAingJuAL4QERsy8xngMuApYDewLSKWAycBZwJ7gGeBy2nd\npL4cGMrMZkQciogzgH3AJcCt09UyMnJwuikzMjjYYHi4OeWcZvMQMH/BMDY2Pm1NnWbSx0KwGPpY\nDD2AffSbbvVxtGA5pk8fdfgUcHd1I/lF4KHMnIiI7cAuWpeZtmTm4YjYAdwXEUPAIeCaah03AA/S\nuhT1RGbu7mJ9kqRpHHMoZOYH255eMMnyncDOjrHXgasmmfscrU8qSZJ6wB9ekyQVhoIkqTAUJEmF\noSBJKgwFSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkSYWhIEkqDAVJUmEoSJIKQ0GSVBgKkqTC\nUJAkFYaCJKkwFCRJhaEgSSoMBUlSYShIkgpDQZJUGAqSpGKgzosiYgD4MnA6sBzYBnwHuBcYB/Zk\n5uZq7nXA9cCbwLbMfCQiTgQeAE4FRoFrM/PliDgHuKua+2Rm3la/NUnSbNU9U/gY8KPM3ABcCnwR\nuBPYkpkbgaURcWVEnAbcBJxbzbs9Ik4AbgReqF5/P7C1Wu8O4OrMPB9YFxFr6zYmSZq9uqHw57x9\nIF8GjAFnZeZQNfYosAk4G9iVmWOZOQrsBdYC64HH2uZeFBENYHlm7qvGHwcurlmfJKmGWqGQmQcz\n87XqQP4V4LPAkrYpTWAV0AAOtI2/CqzuGG+2jY12rGN1nfokSfXUuqcAEBHvBv4C+GJm/llE/Oe2\nxQ1gP62D/KqO8ZFqvNExtznJ3P3T1bFmzUoGBpbVbePHDA42plzeaKzoynZmamBg6bQ1TabOa/rR\nYuhjMfQA9tFv5rKPujeaT6N1eWdzZv7vavhbEbEhM58BLgOeAnYD2yJiOXAScCawB3gWuBx4vvo6\nlJnNiDgUEWcA+4BLgFunq2Vk5GCdFn7C4GCD4eHmlHOazUPA/AXD2Nj4tDV1mkkfC8Fi6GMx9AD2\n0W+61cfRgqXumcJngHcBWyPiFmAC+ATwh9WN5BeBhzJzIiK2A7toXV7akpmHI2IHcF9EDAGHgGuq\n9d4APEjrstYTmbm7Zn2SpBpqhUJm/hbwW5MsumCSuTuBnR1jrwNXTTL3OVqfVJIk9YA/vCZJKgwF\nSVJhKEiSCkNBklQYCpKkwlCQJBWGgiSpMBQkSYWhIEkqDAVJUmEoSJIKQ0GSVBgKkqTCUJAkFYaC\nJKkwFCRJhaEgSSoMBUlSUfdvNEvH7MiRI+zb99Kky0ZG3skrr7za1e2dfvrPs2zZsq6uU1psDAX1\nzL59L/GJL/w1K1efOufbOnjgn/mDm3+Z9773X875tqSFzFBQT61cfSrvXPMzvS5DUsVQUDHV5Zy5\n8Pd///1525akmTEUVMzn5RyAl//fi5zys/9qXrY1MT4+7yF08slr53V7Ujf0XShExBLgS8Ba4A3g\nNzJz/t6+Hufm83LOwQM/nJftALzeHOaO//4jVq7+x3nZ3mv7/4nf+ffDrF49OC/bA2+kqzv6LhSA\nXwFWZOZ5EbEOuLMak47JfAfeLf/16/N21uWNdHVLP4bCeuAxgMz8RkR8oMf19MzE+Dh/93d7Z/Wa\nY/kop9f4u2s+Q2guL48d7XvKM5PFqR9DYRVwoO35WEQszczxXhXUK681X+H6rfdw4jtPnpftHfjh\nS7zrp35hXrYF8HrzFWDJottWL7b3yj8kn7v7O/P2vfLGq6/w29dt4j3v+bl52V43zMXPvnRTv5zl\n9WMojAKNtudTBsKaNSsZGOjOu5XBwcaUy6+84kKWPznUlW3NxMoPnMe69//ivG3vyJEjAPP27u/I\nkQvnbXvzua3jYXutbR5ZcGcKJ5/8zl6XcFTTHX/qzp2tfgyFrwFXAA9FxDnAt6eaPDJysCsbHRxs\nMDzcnHLOSSeu5t98+IqubG+uzKSPhWAx9LEYegD7mC8zra1bfRwtWPoxFB4GNkXE16rnH+9lMZJ0\nPOm7UMjMCeDGXtchSccjf0uqJKkwFCRJhaEgSSoMBUlSYShIkgpDQZJUGAqSpMJQkCQVhoIkqTAU\nJEmFoSBJKgwFSVJhKEiSCkNBklQYCpKkYsnExESva5Ak9QnPFCRJhaEgSSoMBUlSYShIkgpDQZJU\nGAqSpGKg1wX0UkQsAb4ErAXeAH4jM1/qbVX1RMQ3gQPV0/+bmf+ul/XMVkSsAz6fmRdGxHuBe4Fx\nYE9mbu5pcbPQ0cf7gK8C360W78jMr/SuuulFxADwZeB0YDmwDfgOC2x/HKWPH7Dw9sdS4G4gaP3/\nvwE4xBzuj+P9TOFXgBWZeR7wGeDOHtdTS0SsAMjMD1b/LbRAuJnWN/6KauhOYEtmbgSWRsSVPStu\nFibp4/3AHW37pa8PQJWPAT/KzA3ApcAXWZj7o72Py2j1cRYLb398GJjIzPXAVuB3meP9cbyHwnrg\nMYDM/Abwgd6WU9ta4B0R8XhE/M/q3epC8j3gI23P35+ZQ9XjR4GL57+kWn6iD+BDEfF0RNwTEe/o\nUV2z8ee0Dj4Ay4Ax4KwFuD/a+1gKvElrf1yxkPZHZv4VcH319OeAEeZ4fxzvobCKty+5AIxVp2sL\nzUHgC5l5CXAj8KcLqY/MfJjWwectS9oeN4HV81tRPZP08Q3g5uod3UvArb2oazYy82BmvhYRDeAr\nwGdZgPtjkj5+G3gO+NRC2h8AmTkeEfcC24EHmeP9sWAOHHNkFGi0PV+ameO9KuYYfBf4U4DM3Au8\nDPxUTys6Nu37oAHs71Uhx+gvM/Nb1eOHgff1spiZioh3A08B92Xmn7FA98ckfSzI/QGQmb8G/AJw\nD3BS26Ku74/jPRS+BlwOEBHnAN/ubTm1/TpwB0BE/DStb5R/7GlFx+ZvImJD9fgyYGiqyX3s8Yh4\n65LkRcA3e1nMTETEacDjwH/IzPuq4W8ttP1xlD4W4v74WER8unr6BnAEeD4iNlZjXd8fx/Wnj2i9\nW9gUEV+rnn+8l8Ucg53An0TEEK13db++QM943vIp4O6IOAF4EXiox/XUdSPwhxFxGPgn3r423M8+\nA7wL2BoRtwATwCdo9bGQ9sdkfXwSuGuB7Y+/oPVv+2lax+vfBP4WuGeu9oe/JVWSVBzvl48kSW0M\nBUlSYShIkgpDQZJUGAqSpMJQkCQVhoIkqTAUJEnF/wf5w+Eqx9Yv5AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11403e7b8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plasmid_kmers_full = kmers(plasmid_reads, 33)\n",
    "\n",
    "plot_spectrum(plasmid_kmers_full)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 4: &nbsp; de Bruijn Graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next part of the project is to write a Python function that creates a de Bruijn graph from a set of reads.  The arguments passed to this function are the same as those passed to `kmers`: a list of reads and a value for _k_.\n",
    "\n",
    "The function should return two objects:\n",
    "* a sparse directed graph where the vertices are the (_k_-1)-mers in the reads and the edges are _k_-mers\n",
    "* a list of all the vertex strings\n",
    "\n",
    "The easiest way to build a graph for this project is to use the SparseGraph class from the APGL module.  I **_strongly_** recommend you build yourself a \"sandbox\" to play with this class before working on this project:  create a new notebook, import APGL, create some graphs, learn how to add and remove edges, _etc_.  (Hint: brush up on set operations, as well, and in particular learn how to use \"set comprehension\" to create a set.)\n",
    "\n",
    "Here is a suggested outline for your function:\n",
    "* since we need to know how many nodes are in the graph when the SparseGraph object is created use a \"two-pass\" algorithm\n",
    "* first create the list of vertex strings:\n",
    "    * create an empty set\n",
    "    * iterate over all the _k_-mers, and add the prefix (first _k_-1 letters) to the set\n",
    "    * build the vertex list from the set\n",
    "* you'll want a dictionary that associates a node name with the corresponding node number (since graph nodes in APGL are integers)\n",
    "* iterate over the _k_-mers again, this time adding edges to the graph:\n",
    "    * for each _k_-mer, add an edge $ u \\rightarrow v $ where $u$ is the node number for the prefix and $v$ is the node number for the suffix (where \"prefix\" and \"suffix\" refer the to first _k_-1 and last _k_-1 characters)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the de Bruijn graph for the \"tiny\" genome:\n",
    "```\n",
    ">>> tiny_graph, tiny_vertices = deBruijn(tiny_reads, 3)\n",
    "\n",
    ">>> tiny_vertices\n",
    "['CG', 'CA', 'GT', 'AA', 'GG', 'GC', 'AT', 'TG']\n",
    "\n",
    ">>> tiny_graph.getAllEdges()\n",
    "array([[0, 2],\n",
    "       [1, 3],\n",
    "       [2, 7],\n",
    "       [3, 6],\n",
    "       [4, 5],\n",
    "       [5, 0],\n",
    "       [5, 1],\n",
    "       [6, 7],\n",
    "       [7, 4],\n",
    "       [7, 5]], dtype=int32)\n",
    "```\n",
    "Your vertex list might have a different order, but you should see the same strings.  Verify all the edges from the Compeau paper are in the graph.  For example, in this list there are edges `[7, 4]` and `[7, 5]`.  Looking at the vertex list we can see these are the edges `\"TG\" ⟼ \"GG\"` and `\"TG\" ⟼ \"GC\"` in the figure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Documentation and Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "87f5569a066fd224f8d207ee0610cb92",
     "grade": true,
     "grade_id": "de_bruijn_doc",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "source": [
    "use kmers function to create a list of kmers, then using set comprehension find all the prefixes/vertices and then make a list from that. Create a directed SparseGraph object and set vertices to relate each prefix to an int index than can be called later. Then, iterate over kmers again to add the edges to the graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "nbgrader": {
     "checksum": "1247111636b02c1220562aa6dfde7156",
     "grade": true,
     "grade_id": "de_bruijn",
     "locked": false,
     "points": 10,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "def deBruijn(reads, k):\n",
    "    kmers_count = kmers(reads, k)\n",
    "    \n",
    "    vertices_set = {i[:k-1] for i in kmers_count}\n",
    "    vertices_list = list(vertices_set)\n",
    "    \n",
    "    graph = SparseGraph(len(vertices_list), undirected = False)\n",
    "    graph.setVertices(range(len(vertices_list)), vertices_list)\n",
    "    \n",
    "    str_to_int = {graph.getVertex(i): i for i in range(graph.size)}\n",
    "    \n",
    "    for i in kmers_count:\n",
    "        prefix = str_to_int[i[-k+1:]]\n",
    "        suffix = str_to_int[i[:k-1]]\n",
    "        graph.addEdge(suffix, prefix)\n",
    "        \n",
    "    return graph, vertices_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 6],\n",
       "       [1, 7],\n",
       "       [2, 0],\n",
       "       [2, 3],\n",
       "       [3, 1],\n",
       "       [4, 2],\n",
       "       [5, 2],\n",
       "       [5, 4],\n",
       "       [6, 5],\n",
       "       [7, 5]], dtype=int32)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiny_graph.getAllEdges()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CG', 'AA', 'GC', 'CA', 'GG', 'TG', 'GT', 'AT']"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tiny_graph.getVertices(range(len(tiny_vertices)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AA': 1, 'AT': 7, 'CA': 3, 'CG': 0, 'GC': 2, 'GG': 4, 'GT': 6, 'TG': 5}"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{tiny_graph.getVertex(i): i for i in range(tiny_graph.size)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Autograder Tests "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The notebook grader will run these tests in the order shown.  You are welcome to run these tests, too, but be careful to run them in order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hrnmy/anaconda3/lib/python3.4/site-packages/scipy/sparse/compressed.py:730: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  SparseEfficiencyWarning)\n"
     ]
    }
   ],
   "source": [
    "tiny_graph, tiny_vertices = deBruijn(tiny_reads, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "ca4672a7e56a852de12cce22bdaf865f",
     "grade": true,
     "grade_id": "vertex_list",
     "locked": true,
     "points": 5,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert len(tiny_vertices) == 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "9e547eda7e8ae034eaf6c6daa3b34d0e",
     "grade": true,
     "grade_id": "edge_list",
     "locked": true,
     "points": 5,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert len(tiny_graph.getAllEdges()) == 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 5: &nbsp; Eulerian Path "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final part of the project is the function that traverses a de Bruijn graph to produce an assembled sequence.  Fill in the body of the function named `path`.  It takes a graph and its vertex list as arguments and returns a list of nodes on an Eulerian circuit of the graph.\n",
    "\n",
    "Here is an algorithm that isn't very intuitive but is easy to implement:\n",
    "1. intitialize two empty lists, `circuit` (which will be the path) and `stack` (as its name implies, it will be a stack of node numbers)\n",
    "2. pick any node as a starting place; call it the \"current node\"\n",
    "3. repeat the following steps:\n",
    "   - get a list of the successors (outgoing edges) of the current node\n",
    "   - if the successor list is empty:\n",
    "      * if the stack is empty the algorithm terminates\n",
    "      * otherwise add the current node to the circuit\n",
    "      * pop a node from the stack, and make it the current node\n",
    "   - otherwise (the successor list is not empty):\n",
    "      * choose a succcessor node from the list\n",
    "      * remove the edge from the current node to the sucessor\n",
    "      * push the current node on the stack\n",
    "      * make the successor the current node"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A detail: when you add a node to the circuit, if you append the node, the final list will be in reverse order; either reverse the list before you return it, or instead of appending, insert the node at the front of the circuit.\n",
    "\n",
    "You should see something like this when you look for a path in the example graph (`tiny_graph` and `tiny_vertices` are the object returned from the graph building function):\n",
    "```\n",
    ">>> path(tiny_graph, tiny_vertices)\n",
    "[2, 7, 4, 5, 1, 3, 6, 7, 5, 0]\n",
    "```\n",
    "(but again your node numbers might be different)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Challenge:** &nbsp; Can you alter the algorithm so that instead of returning a list of nodes it returns the string of characters in the assembled genome?\n",
    "```\n",
    ">>> path(tiny_graph, tiny_vertices)\n",
    "'GTGGCAATGC'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Documentation and Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "0d4a6a7e87dd2285998fdc71e7a75001",
     "grade": true,
     "grade_id": "path_doc",
     "locked": false,
     "points": 5,
     "solution": true
    }
   },
   "source": [
    "Using algorithm in notebook, keep while loop while the length of the stack is not equal to the number of edges in the graph, then use SparseGraphs.neighbours to make a list of connecting vertices, and loop through successors if not emtpy and if it is, add current_node to the circuit and set a new current_node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "9e4360ab740fd6b49fc515190c1ca7f1",
     "grade": true,
     "grade_id": "path",
     "locked": false,
     "points": 10,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "def path(graph, vertices, start = 0):\n",
    "    \n",
    "    circuit = []\n",
    "    stack = []\n",
    "    \n",
    "    current_node = start\n",
    "    \n",
    "    while len(stack) != graph.getNumEdges():\n",
    "        successors = list(graph.neighbours(current_node))\n",
    "\n",
    "        if not successors:\n",
    "            if not stack: \n",
    "                break\n",
    "            else:\n",
    "                circuit.insert(0, current_node)\n",
    "                current_node = stack.pop(0)\n",
    "        else:\n",
    "            successor = successors[0]\n",
    "            graph.removeEdge(current_node, successor)\n",
    "            stack.append(current_node)\n",
    "            current_node = successor\n",
    "    \n",
    "    string_rep = ''.join([graph.getVertex(i)[:1] for i in circuit])\n",
    "    \n",
    "    return string_rep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hrnmy/anaconda3/lib/python3.4/site-packages/scipy/sparse/compressed.py:730: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  SparseEfficiencyWarning)\n"
     ]
    }
   ],
   "source": [
    "graph_1, vert_1 = deBruijn(tiny_reads, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "test = path(graph_1, vert_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in graph_2.getAllEdges():\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 6],\n",
       "       [1, 7],\n",
       "       [2, 0],\n",
       "       [2, 3],\n",
       "       [3, 1],\n",
       "       [4, 2],\n",
       "       [5, 2],\n",
       "       [5, 4],\n",
       "       [6, 5],\n",
       "       [7, 5]], dtype=int32)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_2.getAllDirEdges()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 4]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(graph_2.neighbours(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Autograder Tests "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The notebook grader will run these tests in the order shown.  You are welcome to run these tests, too, but be careful to run them in order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p = path(tiny_graph, tiny_vertices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "5cfb5ab39b080d81b8ffb2957ddee238",
     "grade": true,
     "grade_id": "path_length",
     "locked": true,
     "points": 5,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "assert len(p) == len(tiny_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "collapsed": false,
    "deletable": false,
    "nbgrader": {
     "checksum": "5fce113982b59076876a4b7557ff1289",
     "grade": true,
     "grade_id": "print_path",
     "locked": true,
     "points": 5,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To Learn More"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are some ideas for optional projects if you want to keep on exploring genome assembly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to create your own set of reads (e.g. to increase the coverage, or to make reads for a different genome) download and run Art.  Here is the command I used to make our data set:\n",
    "```\n",
    "$ art_illumina -i ColE9-J.fasta -l 150 -ss HS25 -f 50 -o ColE9-J\n",
    "```\n",
    "Note: before I ran that command, I copied the first line of the FASTA file and appended it at the end so ART would have a chance to generate \"wraparound\" reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read about a program named KmerGenie (Chikhi and Medvedev, _Bioinformatics_, 2013) that analyzes a set of reads to estimate the best _k_-mer size to pass to an assember.  What does KmerGenie suggest for our plasmid data?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Draw your de Bruijn graph with `graphviz`.  Figure out how to run `graphviz` from a Jupyter notebook and embed the graphs in the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "If you're really daring, see what kind of de Bruijn graph you can make from the plasmid reads. Filter out the rare _k_-mers (why?) and build a graph from the rest.  Can you make repeated calls to the `path` function to get contigs? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy the plasmid reads to ACISS and try to assemble the genome using Velvet or one of the other genome assembly programs installed there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
